core: Fix cors issues with http (close #1486)

SamMorrowDrums opened the pull request:
Adds on-by-default `upgradeToHttps` setting for web

Can override with: 
```
window.RufflePlayer.config = { upgradeToHttps: false }
```

Adds off by default CLI flag that can upgrade to https on desktop too

```
cargo run --package=ruffle_desktop -- --upgrade-to-https
```

## Screenshot with config set to `upgradeToHttps: false`

![Screenshot from 2020-12-11 11 36 40@2x](https://user-images.githubusercontent.com/4811358/101899898-d511e480-3bae-11eb-87a7-e9048f3dbd99.png)

## Screenshot with default `upgradeToHttps: true`

-  console info logs are removed in actual PR

![Screenshot from 2020-12-11 11 31 56@2x](https://user-images.githubusercontent.com/4811358/101899900-d5aa7b00-3bae-11eb-970a-8bc5af51cdb7.png)

SamMorrowDrums commented:
Regarding `Add a setting to pattern match and replace URLs` - where would the setting live? Were you thinking an API call when instantiating the browser player?

In the browser context I think it would be best to simply check automatically if it's an https context, and then if it is, set the flag automatically.

SamMorrowDrums commented on web/src/lib.rs:
```diff
@@ -345,6 +348,7 @@ impl Ruffle {
             key_up_callback: None,
             timestamp: None,
             has_focus: false,
+            https_context: protocol == "https:",
```
This isn't hooked up yet, but it's idea for where we could put the https context flag, and then I guess consume it in the player somehow.

Herschel commented on core/src/avm1/activation.rs:
```diff
@@ -1201,9 +1202,21 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
+        if !self.context.https_context || url.is_empty() || !url.contains("://") {
+            Ok(url.to_string())
+        } else {
+            let mut url = Url::parse(url).unwrap();
+            if url.scheme() == "http" {
+                url.set_scheme("https").unwrap();
```
Let's avoid the `unwraps` here; either pass the failure upward with `?` or return the URL as-is on failure.

SamMorrowDrums replied:
I had trouble successfully using ? as I couldn't get map_err to return the right error type. That was the intention.

That said, it seems I might be moving the code from here anyway.

Herschel commented:
Thanks!

* Certainly this should be a setting, the question is whether it should be on by default. I get an icky feeling from automatically changing URLs without an explicit setting, but maybe http -> https is fine and expected.
* This could be covered more generally by a generic URL swap. For example, for web, this could be a Javascript callback that the user can set:
```
config.handleUrl = (url) => {
   url.replace(/^http:/i, "https:");
}
```
The question is: should that be the only method of spoofing, and what is used for HTTP->HTTPS? Or should we also provide a "simpler" `upgradeHttps` setting since that could be a common case?

Code nits:
 * I think it'd be better for the URL swapping code to live inside the `NavigatorBackend`; this allows for platform-specific behavior (such as a JS callback, see above), and also avoids having to fiddle with the UpdateContext. `WebNavigatorBackend` can store `https_context` bool within itself.

SamMorrowDrums commented:
Thanks that's a good suggestion and happy to move to JavaScript land for the callback.

I personally think there's nothing that icky about default upgrading http -> https in situations where it will never succeed because browser policies forbid it outright (especially assuming that legacy content inside flash files will probably not be upgraded), but I guess the soft approach of enabling users to do the change themselves and having a default callback which can be configured by user, and enabled with a flag is ok.

It's certainly the least interventionist approach, which is usually preferable. But I'm not sure if there's much benefit in supporting inevitable failure (unless we really think it's just not our place to automatically intervene).

I also think an upgrade flag should be available on the CLI, as users might legitimately prefer to work with https only with known risk that it might not work.

There's even an argument that it should be on by default unless the user wishes to opt out (as a more secure by default approach), however I think we can have the discussion after getting something optional released in the first place.

Will take a look at Navigator - see what I can come up with.

Herschel commented:
Ok, thought about it some more:
 * Add `upgradeToHttps` setting to the config options that defaults to true. This will automatically upgrade a URL to https when the site itself is https. This is common enough that it should default to true.
 * The above should be handled by the NavigatorBackend; it could provide a method that takes in a Url and returns it possibly modified.
 * Defer to a later PR: An additional `handleUrl` setting (unsure of the name) that will allow a callback and more fine-grained tuning for URL spoofing. It will run after any `upgradeToHttps` handling.

SamMorrowDrums commented:
Yes, I think I am fully on the right track then. I've almost got next PR
ready.

I also renamed config param to "upgradeHttp" as suggested which I think is
more descriptive.

I have added a default false param for the desktop version, so you a user
can pass CLI flag for same behaviour if they wishes to. Will probably get
it out for tomorrow.

Dropped the ball missing the Navigator as a place to put the code. Much
simpler.

On Thu, 10 Dec 2020, 21:24 Mike Welsh, <notifications@github.com> wrote:

> Ok, thought about it some more:
>
>    - Add upgradeToHttps setting to the config options that defaults to
>    true. This will automatically upgrade a URL to https when the site itself
>    is https. This is common enough that it should default to true.
>    - The above should be handled by the NavigatorBackend; it could
>    provide a method that takes in a Url and returns it possibly modified.
>    - Defer to a later PR: An additional handleUrl setting (unsure of the
>    name) that will allow a callback and more fine-grained tuning for URL
>    spoofing. It will run after any upgradeToHttps handling.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ruffle-rs/ruffle/pull/1897#issuecomment-742808966>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEWUXWOSNC2WIGWMNKVRBTSUE4CNANCNFSM4UTQRJVA>
> .
>

SamMorrowDrums commented:
So yes will do upgradeToHttps and and things as suggested.

On Thu, 10 Dec 2020, 22:26 Sam Morrow, <info@sammorrowdrums.com> wrote:

> Yes, I think I am fully on the right track then. I've almost got next PR
> ready.
>
> I also renamed config param to "upgradeHttp" as suggested which I think is
> more descriptive.
>
> I have added a default false param for the desktop version, so you a user
> can pass CLI flag for same behaviour if they wishes to. Will probably get
> it out for tomorrow.
>
> Dropped the ball missing the Navigator as a place to put the code. Much
> simpler.
>
> On Thu, 10 Dec 2020, 21:24 Mike Welsh, <notifications@github.com> wrote:
>
>> Ok, thought about it some more:
>>
>>    - Add upgradeToHttps setting to the config options that defaults to
>>    true. This will automatically upgrade a URL to https when the site itself
>>    is https. This is common enough that it should default to true.
>>    - The above should be handled by the NavigatorBackend; it could
>>    provide a method that takes in a Url and returns it possibly modified.
>>    - Defer to a later PR: An additional handleUrl setting (unsure of the
>>    name) that will allow a callback and more fine-grained tuning for URL
>>    spoofing. It will run after any upgradeToHttps handling.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ruffle-rs/ruffle/pull/1897#issuecomment-742808966>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABEWUXWOSNC2WIGWMNKVRBTSUE4CNANCNFSM4UTQRJVA>
>> .
>>
>

SamMorrowDrums commented:
OK @Herschel I think this is ready for review

SamMorrowDrums commented:
One additional note: I tried to make it a generic type / trait implementation defined type of either `Url` or `str`, for the trait - but then you have to specify the concrete type when putting a `NavigatorBackend` into a `Box` etc. which obviously defeats the object of the trait, alternatively we could use an enum, with a `str` and `Url` and then the return type would be that enum as well, but that seemed like a bit of a hack.

I think perhaps the solution might be to remove the fn from the `NavigatorBackend` trait, and instead have definitions only in their own struct implementations, perhaps that's OK in this case - but I will have to take advice on that.

On another note the function for desktop got more complex calling self inside the async block, so while I would have preferred to remove the extra `Url::parse`, I couldn't find a good way to do it - without adding additional complexity (unless I could resolve the above).

Toad06 commented:
Thanks for this PR, I've tested it and works great (happy to see that [Soap Bubble II](https://www.newgrounds.com/portal/view/347388) is now playable). 👍

The only little thing is that when the site itself is `http:`, requests are still upgraded to `https:`. I don't think that's intended and it seems this can be fixed easily in `ruffle-player.ts`: `config.upgradeToHttps !== false && window.location.protocol === "https:"`

SamMorrowDrums commented:
@Toad06 thanks for feedback, @Herschel said actually I should  `Add upgradeToHttps setting to the config options that defaults to true` - as a good default, and then if stuff breaks because it's not able to get https, then it can be configured for that site.

That said I'm sure he'll see your comment, and it's easy to change if we decide to base on context.

SamMorrowDrums commented:
I think @Toad06 is right really, and I should consider going back to context specific https upgrade by default - otherwise there'll be some edge case breakage that probably isn't the responsibility of the player.

Herschel commented:
I agree with @Toad06, on web, `upgradeToHttps` should only take effect if the site itself is hosted on https.  `upgradeToHttps` can still default to true, it will just be ignored in these situations.
(On desktop, it could always take effect regardless).

SamMorrowDrums commented:
OK should be correct behaviour now. There is the small case of opening files with remote links on localhost, but I guess if you have the swf, desktop player is the way to go anyway. PR should help a bunch of cases at least now :-)

Herschel commented:
Closed in favor of #1930. Thank you!

Full desktop behavior is a whole other TODO (have to inspect and obey the sandbox setting of the SWF), so bumping all desktop requests to https is acceptable for now.
